archive/frames/views.py

```diff
@@ -163,83 +168,174 @@ def zip(self, request):
-        query_filters = {}
-        for k in FIELD_MAPPING.keys():
-            if k in request.GET:
-                query_filters[FIELD_MAPPING[k]] = request.GET[k]
-        for k in fields:
-            if k in request.GET:
-                query_filters[k] = request.GET[k]
-        if 'start' in request.GET:
-            query_filters['start'] = parse(request.GET['start']).replace(tzinfo=UTC, second=0, microsecond=0)
-        if 'end' in request.GET:
-            query_filters['end'] = parse(request.GET['end']).replace(tzinfo=UTC, second=0, microsecond=0)
-        cache_hash = blake2s(repr(frozenset(list(query_filters.items()) + [aggregate_field])).encode()).hexdigest()
-        response_dict = cache.get(cache_hash)
-        if not response_dict:
-            qs = Frame.objects.all()
-            qs = DjangoFilterBackend().filter_queryset(request, qs, view=self)
-            sites = self._get_aggregate_values(qs, query_filters, 'site_id', aggregate_field)
-            telescopes = self._get_aggregate_values(qs, query_filters, 'telescope_id', aggregate_field)
-            filters = self._get_aggregate_values(qs, query_filters, 'primary_optical_element', aggregate_field)
-            instruments = self._get_aggregate_values(qs, query_filters, 'instrument_id', aggregate_field)
-            obstypes = self._get_aggregate_values(qs, query_filters, 'configuration_type', aggregate_field)
-            proposals = self._get_aggregate_values(qs.filter(public_date__lte=timezone.now()), query_filters, 'proposal_id', aggregate_field)
-            response_dict = {
-                'sites': sites,
-                'telescopes': telescopes,
-                'filters': filters,
-                'instruments': instruments,
-                'obstypes': obstypes,
-                'proposals': proposals
-            }
-            cache.set(cache_hash, response_dict, 60 * 60)
+
+        cache_key_elms = [
+            start,
+            end,
+            public,
+            site_id,
+            telescope_id,
+            primary_optical_element,
+            instrument_id,
+            configuration_type,
+            proposal_id,
+        ]
+
+        # unauthenticated users share the same cache space
+        # authenticated users each get their own cache space
+        if request.user.is_authenticated:
+            cache_key_elms.append(request.user.id)
+
+        cache_key = blake2b("".join(map(str, cache_key_elms)).encode("utf-8")).hexdigest()
+
+        response_dict = cache.get(cache_key)
+
+        if response_dict:
+            response_serializer = self.get_response_serializer(response_dict)
+            return Response(response_serializer.data)
+
+        frames = Frame.objects.all()
+
+        frames = frames.filter(observation_date__gte=start)
+
+        frames = frames.filter(observation_date__lt=end)
+
+        if not public:
+            logger.info("excluding public data")
+            frames = frames.exclude(public_date__lt=Now())
```

Comment:
one danger of this, which I just realized, is that this will exclude a users own data if it happens to be public, like if its older than 1 year in the past... I realize this is how the filter is already coded, so it's already a problem if they uncheck the include public option. Hopefully this isn't an issue for most users since they are looking at data in the last two semesters, but it could be confusing to some.

Reply:
Yeah, I guess it all depends on your interpenetration of public. Maybe I could do something like this for authenticated users:
```
 frames = frames.exclude(
                    Q(public_date__lt=Now()) & ~Q(proposal_id__in=user_proposals)
                )
```

Only exclude public frames that don't have an ID that's in proposal set.